File: python/VCFSmallFillable.py
from __future__ import annotations

# coding: utf-8
# VCFSmallFillable.py

from itertools import *
from collections import defaultdict

from typing import Optional

from VCF import VCFSmall
from GenoRecord import GenoRecord
from VCFGeno import VCFGeno
from VCFFamily import *
from VCFImpFamilyRecord import FillType, VCFImpFamilyRecord
from VCFFillableRecord import VCFFillableRecord
from VCFFillable import *
from VCFHeteroHomo import *
from group import Groups
from RecordSet import RecordSet, RecordSetSmall
from Genotype import Genotype
from option import *
from common import *


#################### VCFSmallFillable ####################

class VCFSmallFillable(VCFFillable):
	def __init__(self, samples: list[str],
							records: list[VCFFillableRecord], vcf: VCFSmall):
		VCFFillable.__init__(self, samples, records, vcf)
	
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	def get_family_record(self, i: int) -> VCFFamilyRecord:
		return self.records[i]
	
	def modify(self, is_phased_changable: bool) -> None:
		# FillTypeでrecordを分ける
		groups = Groups.create(self.records)
		for record_set in groups.generate_record_sets_small():
			record_set.determine_parents_phasing()
			self.impute_core(record_set)
		
		for i, record in enumerate(self.records):
			# 家系ごとで./.にしたGenotypeを補完
			if record.type == FillType.MAT:
				self.impute_NA_mat(i)
			elif record.type == FillType.PAT:
				self.impute_NA_pat(i)
			elif record.type in (FillType.IMPUTABLE, FillType.UNABLE):
				self.impute_others(i)
	
	# 家系ごとで./.にしたGenotypeを補完
	def impute_NA_pat(self, i: int) -> None:
		record = self.records[i]
		for k in range(2, len(record.geno)):
			if record.is_NA(k):
				self.impute_NA_pat_each(i, k)
	
	def remove_parents(self) -> VCFGeno:
		samples: list[str] = self.samples[2:]
		records: list[GenoRecord] = []
		for record in self.records:
			geno = record.geno[2:]
			r = GenoRecord(record.pos, geno)
			records.append(r)
		return VCFGeno(samples, records, self.vcf)


__all__ = ['VCFSmallFillable']

File: src/VCFSmallFillable.cpp
#include <limits>
#include <memory>
#include <cassert>
#include "../include/VCFSmallFillable.h"
#include "../include/VCFHeteroHomo.h"
#include "../include/group.h"

#include "../include/option.h"
#include "../include/common.h"

using namespace std;


//////////////////// VCFSmallFillable ////////////////////

void VCFSmallFillable::phase_in_thread(void *config) {
	auto	*c = static_cast<const ConfigThreadPhase *>(config);
	const auto&	record_sets = c->record_sets;
	const size_t	n = record_sets.size();
	for(size_t i = c->first; i < n; i += c->num_threads) {
		record_sets[i]->impute(true);
	}
}

void VCFSmallFillable::modify(int T) {
	const Groups	*groups = Groups::create(records);
	const auto	record_sets = groups->create_record_sets_small();
	
	vector<ConfigThreadPhase *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThreadPhase(i, T, record_sets, this);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
			(void *(*)(void *))&phase_in_thread, (void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		phase_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
	
	delete groups;
	Common::delete_all(record_sets);
	
	for(size_t i = 0U; i < records.size(); ++i) {
		auto	*record = records[i];
		if(record->is_mat_type())
			this->impute_NA_mat(i);
		else if(record->is_pat_type())
			this->impute_NA_pat(i);
		else if(record->is_fillable_type())
			this->impute_others(i);
	}
	
	for(auto p = records.begin(); p != records.end(); ++p)
		(*p)->fill_PGT();
}

const RecordSet *VCFSmallFillable::create_recordset(
										size_t i, size_t c, bool is_mat) const {
	auto	*record = records[i];
	auto	*prev_record = find_prev_same_type_record(i, c);
	auto	*next_record = find_next_same_type_record(i, c);
	if(is_mat)
		return new RecordSetSmall(record, prev_record, next_record, NULL, NULL);
	else
		return new RecordSetSmall(record, NULL, NULL, prev_record, next_record);
}
